### PR TITLE
Check before insert

### PR DESCRIPTION
**Pull request by shawn-hurley**
**Description of the change:**

* adding sanity check before adding a bundle to check it the package/csv/version is already provided or if the exact bundle is already added.
* exit code's now can be used to retrieve these specific error scenarios. 
* This is targeted at index add, but anything that uses the directory loader will get this change.

**Motivation for the change:**

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

File: pkg/lib/indexer/indexer.go [hunk 1]
```diff
@@ -85,6 +85,7 @@ func (i ImageIndexer) AddToIndex(request AddToIndexRequest) error {
 	// Add the bundles to the registry
 	err = i.RegistryAdder.AddToRegistry(addToRegistryReq)
 	if err != nil {
+		fmt.Printf("add to registry error: %#v", err)
```
Comment (gallettilance) [inline pkg/lib/indexer/indexer.go:88]: How about:
```suggestion
		return fmt.Errorf("add to registry error: %#v", err)
```

File: pkg/lib/registry/registry.go [hunk 2]
```diff
@@ -96,16 +95,17 @@ func (r RegistryUpdater) AddToRegistry(request AddToRegistryRequest) error {
 	}
 
 	if err := populate(context.TODO(), dbLoader, graphLoader, dbQuerier, reg, simpleRefs, request.Mode); err != nil {
-		err = fmt.Errorf("error loading bundle from image: %s", err)
+		r.Logger.Debugf("unable to populate database: %s", err)
```
Comment (gallettilance) [inline pkg/lib/registry/registry.go:98]: Not sure I understand why this change is needed

File: pkg/registry/populator.go [hunk 3]
```diff
@@ -63,7 +63,54 @@ func (i *DirectoryPopulator) Populate(mode Mode) error {
...
+				errs = append(errs, err)
+				return utilerrors.NewAggregate(errs)
+			}
+			if bundle != nil && !isExactImage {
+				// raise error that this package + channel + csv combo is already in the db
+				errs = append(errs, BundleAlreadyInDatabaseError{ErrorString: "Bundle already added that provides package and csv"})
+				continue
+			}
+			if bundle != nil && isExactImage {
+				// raise error that this is already in the db
```
Comment (gallettilance) [inline pkg/registry/populator.go:97]:
```suggestion
				// raise error that this bundle and channel is already in the db
```

File: pkg/registry/populator.go [hunk 4]
```diff
@@ -63,7 +63,54 @@ func (i *DirectoryPopulator) Populate(mode Mode) error {
...
+				errs = append(errs, BundleAlreadyAddedError{ErrorString: "Bundle already added"})
```
Comment (gallettilance) [inline pkg/registry/populator.go:98]:
```suggestion
				errs = append(errs, BundleAlreadyAddedError{ErrorString: fmt.Sprintf("Bundle %s already exists for channel %s", image.bundle.csv.GetName(), channel)})
```

File: cmd/opm/main.go [hunk 5]
```diff
@@ -35,6 +37,16 @@ func main() {
 	}
 
 	if err := rootCmd.Execute(); err != nil {
+		if agg, ok := err.(utilerrors.Aggregate); ok {
+			for _, e := range agg.Errors() {
+				if _, ok := e.(registrylib.BundleAlreadyAddedError); ok {
+					os.Exit(2)
```
Comment (gallettilance) [inline cmd/opm/main.go:43]: neat

File: pkg/registry/types.go [hunk 6]
```diff
@@ -14,6 +14,24 @@ var (
 	ErrPackageNotInDatabase = errors.New("Package not in database")
 )
 
+// BundleAlreadyAddedError is an error that describes a bundle is already added
```
Comment (kevinrizza) [inline pkg/registry/types.go:17]: What is the semantic difference between these two error types? Is there value in differentiating between that exact bundle was already added vs that specific csv was already added? Or is this doing something more subtle than what I'm reading?

Either way, the comments/naming here is confusing.

File: pkg/registry/populator.go [hunk 7]
```diff
@@ -63,7 +63,55 @@ func (i *DirectoryPopulator) Populate(mode Mode) error {
...
+				isExactImage = true
+			}
+		}
+		for _, channel := range image.bundle.Channels {
+			bundle, err := i.querier.GetBundle(context.TODO(), image.bundle.Package, channel, image.bundle.csv.GetName())
+			if err != nil {
+				// Assume that if we can not find a bundle for the package, channel and or CSV Name that this is safe to add
+				continue
+			}
+			if bundle != nil && !isExactImage {
```
Comment (kevinrizza) [inline pkg/registry/populator.go:92]: This wouldn't really need to be per image though, right? Even in disparate channels, you can't add the same bundle to the registry twice.

File: pkg/registry/populator.go [hunk 8]
```diff
@@ -63,7 +63,55 @@ func (i *DirectoryPopulator) Populate(mode Mode) error {
 	return nil
 }
 
+func (i *DirectoryPopulator) globalSanityCheck(imagesToAdd []*ImageInput) error {
```
Comment (ecordell) [inline pkg/registry/populator.go:66]: Just to clarify - these errors should already be returned by the normal loader, and this PR is specifically to get different return codes for some cases?

This seems fine for now, but IMO we should look at doing something similar to what we plan to do for bundle validation, and output a set of error codes (i.e. in json).

File: pkg/registry/populator.go [hunk 9]
```diff
@@ -63,7 +63,55 @@ func (i *DirectoryPopulator) Populate(mode Mode) error {
...
+	var errs []error
+	images := make(map[string]bool)
+	for _, image := range imagesToAdd {
+		images[image.bundle.BundleImage] = true
+	}
+
+	for _, image := range imagesToAdd {
```
Comment (ecordell) [inline pkg/registry/populator.go:73]: this is making a lot of queries that could be wrapped up into larger sql queries, but it's not worth doing that now, IMO

File: pkg/registry/populator.go (hunk 8 above)
Comment (shawn-hurley) [reply pkg/registry/populator.go:66]: +1 on adding that in the future.

File: pkg/registry/populator_test.go [hunk 10]
```diff
@@ -426,6 +505,18 @@ func TestImageLoading(t *testing.T) {
 	}
 }
 
+func checkAggErr(aggErr, wantErr error) bool {
+	if a, ok := aggErr.(utilerrors.Aggregate); ok {
+		for _, e := range a.Errors() {
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply pkg/registry/populator_test.go:510]: I don't think that is true, I am looking through each error to see if one of them is the error that I want, not that all the errors are the one that I want.

File: pkg/registry/types.go (hunk 6 above)
Comment (shawn-hurley) [reply pkg/registry/types.go:17]: the difference is

A bundle with a different pull spec providing the same package and csv in the same channels.

vs

The exact same bundle pull spec providing the same things.

WDYT of `PackageVersionAlreadyAdded` for the first case and `BundleAlreadyAdded` for the second?

File: pkg/registry/populator.go (hunk 9 above)
Comment (kevinrizza) [reply pkg/registry/populator.go:73]: Agreed, we should probably just load the entire package in once, but it doesn't need to happen now.

File: pkg/registry/populator.go (hunk 7 above)
Comment (shawn-hurley) [reply pkg/registry/populator.go:92]: meaning we should exit early once we find the same bundle path on line 83?

File: pkg/registry/populator.go (hunk 9 above)
Comment (shawn-hurley) [reply pkg/registry/populator.go:73]: +1

File: pkg/registry/types.go (hunk 6 above)
Comment (kevinrizza) [reply pkg/registry/types.go:17]: I like `PackageVersionAlreadyAdded`, but for the second what about `BundleImageAlreadyAdded`?

File: pkg/registry/populator.go (hunk 7 above)
Comment (kevinrizza) [reply pkg/registry/populator.go:92]: Yeah, I think if `isExactImage` is set to true you should just return the error for that entire bundle and move to the next one, no need to iterate over the channels at all.

File: pkg/registry/types.go (hunk 6 above)
Comment (shawn-hurley) [reply pkg/registry/types.go:17]: +1 on BundleImageAlreadyAdded changes coming

Comment (shawn-hurley) [reply pkg/registry/types.go:17]: Adding `Err` on them as well

File: pkg/registry/populator.go (hunk 7 above)
Comment (shawn-hurley) [reply pkg/registry/populator.go:92]: Do you think that we could also break out of the channel check loop once we know that one is failing?

File: pkg/registry/populator_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (ecordell) [reply pkg/registry/populator_test.go:510]: ah, I misread. `require.Contains` would work, but it's not a big deal.

File: pkg/registry/populator.go [hunk 11]
```diff
@@ -63,7 +63,50 @@ func (i *DirectoryPopulator) Populate(mode Mode) error {
 	return nil
 }
 
+func (i *DirectoryPopulator) globalSanityCheck(imagesToAdd []*ImageInput) error {
+	var errs []error
+	images := make(map[string]bool)
+	for _, image := range imagesToAdd {
+		images[image.bundle.BundleImage] = true
+	}
```
Comment (estroz) [inline pkg/registry/populator.go:71]: Nit:
```suggestion
	images := make(map[string]struct{})
	for _, image := range imagesToAdd {
		images[image.bundle.BundleImage] = struct{}{}
	}
```

File: pkg/registry/types.go [hunk 12]
```diff
@@ -14,6 +14,24 @@ var (
 	ErrPackageNotInDatabase = errors.New("Package not in database")
 )
 
+// BundleImageAlreadyAddedErr is an error that describes a bundle is already added
```
Comment (estroz) [inline pkg/registry/types.go:17]: How about adding an `ExitCoder` interface to simplify/allow extension of `opm` error code usage: 
```go
type ExitCoder interface {
	ExitCode() int
}

type BundleImageAlreadyAddedErr struct {
	ErrorString string
}

...

func (e BundleImageAlreadyAddedErr) ExitCode() int {
	return 2
}
```

File: pkg/registry/populator_test.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply pkg/registry/populator_test.go:510]: hum looks that this actually compares error strings and not error types:

```
    TestImageLoading/AddExactBundleAlreadyExists: populator_test.go:509:
                Error Trace:    populator_test.go:509
                                                        populator_test.go:489
                Error:          "[Bundle quay.io/prometheus/operator:0.14.0 already exists Bundle already added that provides package and csv]" does not contain ""
                Test:           TestImageLoading/AddExactBundleAlreadyExists
```

Just an FYI

File: pkg/registry/populator.go (hunk 11 above)
Comment (kevinrizza) [reply pkg/registry/populator.go:71]: +1

File: pkg/registry/types.go (hunk 12 above)
Comment (estroz) [reply pkg/registry/types.go:17]: Disregard this comment, the exit code is bringing CLI problems into a non-CLI-specific library.
